Investigate TS 7.0 ( `tsgo` ) type issues

Tried adding a new typecheck job with `tsgo`, and got these errors:

```
Version 7.0.0-dev.20251207.1
Error: src/query/react/buildHooks.ts(1823,13): error TS2769: No overload matches this call.
  The last overload gave the following error.
    Argument of type '((currentState: ({ status: QueryStatus.uninitialized; originalArgs?: undefined; data?: undefined; error?: undefined; requestId?: undefined; endpointName?: string | undefined; startedTimeStamp?: undefined; fulfilledTimeStamp?: undefined; } & { ...; }) | ({ ...; } & { ...; }) | ({ ...; } & { ...; }) | ({ ...; } & { .....' is not assignable to parameter of type '(resultFuncArgs_0: ({ status: QueryStatus.uninitialized; originalArgs?: undefined; data?: undefined; error?: undefined; requestId?: undefined; endpointName?: string | undefined; startedTimeStamp?: undefined; fulfilledTimeStamp?: undefined; } & { ...; }) | ({ ...; } & { ...; }) | ({ ...; } & { ...; }) | ({ ...; } & {...'.
      Type '(currentState: ({ status: QueryStatus.uninitialized; originalArgs?: undefined; data?: undefined; error?: undefined; requestId?: undefined; endpointName?: string | undefined; startedTimeStamp?: undefined; fulfilledTimeStamp?: undefined; } & { ...; } & { ...; } & InfiniteQueryResultFlags) | ({ ...; } & ... 2 more ... ...' is not assignable to type '(resultFuncArgs_0: ({ status: QueryStatus.uninitialized; originalArgs?: undefined; data?: undefined; error?: undefined; requestId?: undefined; endpointName?: string | undefined; startedTimeStamp?: undefined; fulfilledTimeStamp?: undefined; } & { ...; }) | ({ ...; } & { ...; }) | ({ ...; } & { ...; }) | ({ ...; } & {...'.
        Types of parameters 'currentState' and 'resultFuncArgs' are incompatible.
          Type '({ status: QueryStatus.uninitialized; originalArgs?: undefined; data?: undefined; error?: undefined; requestId?: undefined; endpointName?: string | undefined; startedTimeStamp?: undefined; fulfilledTimeStamp?: undefined; } & { ...; }) | ({ ...; } & { ...; }) | ({ ...; } & { ...; }) | ({ ...; } & { ...; })' is not assignable to type '({ status: QueryStatus.uninitialized; originalArgs?: undefined; data?: undefined; error?: undefined; requestId?: undefined; endpointName?: string | undefined; startedTimeStamp?: undefined; fulfilledTimeStamp?: undefined; } & { ...; } & { ...; } & InfiniteQueryResultFlags) | ({ ...; } & ... 2 more ... & InfiniteQuery...'.
            Type '{ status: QueryStatus.uninitialized; originalArgs?: undefined; data?: undefined; error?: undefined; requestId?: undefined; endpointName?: string | undefined; startedTimeStamp?: undefined; fulfilledTimeStamp?: undefined; } & { ...; }' is not assignable to type '({ status: QueryStatus.uninitialized; originalArgs?: undefined; data?: undefined; error?: undefined; requestId?: undefined; endpointName?: string | undefined; startedTimeStamp?: undefined; fulfilledTimeStamp?: undefined; } & { ...; } & { ...; } & InfiniteQueryResultFlags) | ({ ...; } & ... 2 more ... & InfiniteQuery...'.
              Type '{ status: QueryStatus.uninitialized; originalArgs?: undefined; data?: undefined; error?: undefined; requestId?: undefined; endpointName?: string | undefined; startedTimeStamp?: undefined; fulfilledTimeStamp?: undefined; } & { ...; }' is not assignable to type '{ status: QueryStatus.uninitialized; originalArgs?: undefined; data?: undefined; error?: undefined; requestId?: undefined; endpointName?: string | undefined; startedTimeStamp?: undefined; fulfilledTimeStamp?: undefined; } & { ...; } & { ...; } & InfiniteQueryResultFlags'.
                Type '{ status: QueryStatus.uninitialized; originalArgs?: undefined; data?: undefined; error?: undefined; requestId?: undefined; endpointName?: string | undefined; startedTimeStamp?: undefined; fulfilledTimeStamp?: undefined; } & { ...; }' is missing the following properties from type 'InfiniteQueryResultFlags': hasNextPage, hasPreviousPage, isFetchingNextPage, isFetchingPreviousPage, and 2 more.
Error: src/tests/mapBuilders.test-d.ts(56,7): error TS2578: Unused '@ts-expect-error' directive.
Error: src/tests/mapBuilders.test-d.ts(58,9): error TS2769: No overload matches this call.
  The last overload gave the following error.
    Argument of type 'ActionCreatorWithPayload<number, "increment">' is not assignable to parameter of type 'string'.
Error: src/tests/mapBuilders.test-d.ts(67,7): error TS2578: Unused '@ts-expect-error' directive.
Error: src/tests/mapBuilders.test-d.ts(70,9): error TS2769: No overload matches this call.
  The last overload gave the following error.
    Argument of type '(state: number, action: { payload: number; type: "increment"; }) => number' is not assignable to parameter of type 'CaseReducer<number, Action<"decrement">>'.
      Types of parameters 'action' and 'action' are incompatible.
        Property 'payload' is missing in type 'Action<"decrement">' but required in type '{ payload: number; type: "increment"; }'.
Error: Process completed with exit code 2.

```

Reading those, seems like 2-3 different issues:

- `buildHooks.ts:1823`: an issue with the `preSelector` parameter , which could be for standard or infinite queries, and how it ends up dragging in the infinite query result flags types
- `map-builders.test-d.ts:56,67`: the tests for `addCase` have `@ts-expect-error`s, but now it's not erroring on the expected lines
- `map-builders.test-d.ts:58`: some overload errors (possibly "just" that the earlier expected errors have now migrated a couple lines down?)

--- a/.github/workflows/tests.yml
+++ b/.github/workflows/tests.yml
@@ -465,7 +465,6 @@ jobs:
           yarn workspace @examples-type-portability/${{ matrix.example.name }} run test --module ESNext --moduleResolution Node10 --preserveSymLinks --verbatimModuleSyntax false
 
   test-types-native-preview:
-    if: false # Disabled pending TS 7.0 compatibility fixes
     name: 'Test Types: TypeScript Native Preview'
     needs: [build]
     runs-on: ubuntu-latest

--- a/packages/toolkit/src/query/react/buildHooks.ts
+++ b/packages/toolkit/src/query/react/buildHooks.ts
@@ -1856,6 +1856,11 @@ export function buildHooks<Definitions extends EndpointDefinitions>({
               (_: ApiRootState, lastResult: any) => lastResult,
               (_: ApiRootState) => stableArg,
             ],
+            // `preSelector` is a union of the standard/infinite query pre-selectors.
+            // `tsc` attributes the resulting overload error to the `createSelector` call
+            // (suppressed above), but `tsgo` attributes it to this argument line, so it
+            // needs its own directive.
+            // @ts-ignore
             preSelector,
             {
               memoizeOptions: {

--- a/packages/toolkit/src/tests/mapBuilders.test-d.ts
+++ b/packages/toolkit/src/tests/mapBuilders.test-d.ts
@@ -52,22 +52,20 @@ describe('type tests', () => {
         (state, action: ReturnType<typeof increment>) => state,
       )
 
-      // @ts-expect-error
-      builder.addCase(
-        increment,
-        (state, action: ReturnType<typeof decrement>) => state,
-      )
+      // @ts-expect-error `increment` and `decrement` action types are incompatible.
+      // The call is kept on a single line so `tsc` and `tsgo` attribute the error to the same line.
+      // prettier-ignore
+      builder.addCase(increment, (state, action: ReturnType<typeof decrement>) => state)
 
       builder.addCase(
         'increment',
         (state, action: ReturnType<typeof increment>) => state,
       )
 
-      // @ts-expect-error
-      builder.addCase(
-        'decrement',
-        (state, action: ReturnType<typeof increment>) => state,
-      )
+      // @ts-expect-error The reducer is typed for `increment`, but the case is `'decrement'`.
+      // The call is kept on a single line so `tsc` and `tsgo` attribute the error to the same line.
+      // prettier-ignore
+      builder.addCase('decrement', (state, action: ReturnType<typeof increment>) => state)
 
       // action type is inferred
       builder.addMatcher(increment.match, (state, action) => {

